File: dnn-providers/hip-kernel-provider/rocke/platform/Python/rocke/helpers/preshuffle.py
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT

"""Preshuffled-B layout helpers.

CK Tile's preshuffled-B GEMM family (``03_gemm`` Preshuffle pipeline,
``18_flatmm``, the ``38_block_scale_gemm`` ``preshuffleb`` columns)
reorders the B-matrix tiles on the host side into a "tile-major"
layout::

 canonical [N, K] row-major; per-element stride (K, 1)
 preshuffled [N_outer, K_outer, N_inner, K_inner]
 with (N_inner * K_inner) == (block_n * block_k)
 packed contiguously

so the per-K-iter load issues *one* aligned ``buffer_load_dwordx4``
per warp (no scattered ds_write strides on the load side, no
``ds_read_b32`` strides on the consumer side). The downside is a host-
side permutation of the B matrix; the upside is a 1.3-1.8x speedup on
fp8 / bf8 / i4 GEMMs where the standard pipeline is bandwidth-bound on
the consumer-side LDS reads.

This module is the *layout descriptor* + load-distribution helper. It
does not own the per-element load -- that lives in
``CoalescedTileLoader`` / ``AsyncTileLoader`` -- but it does compute
the per-thread byte offset into the preshuffled buffer.

What v1 ships:

* :class:`PreshuffleBSpec`: tile shape + element bytes; computes the
 per-tile element stride into the preshuffled layout.
* :func:`emit_preshuffleb_offset`: per-lane byte offset for one
 ``(n_tile, k_tile, n_in_tile, k_in_tile)`` quad.
* :func:`host_preshuffle_layout`: pure-Python description of the
 layout for the host-side permutation pass (returns a numpy-style
 shape + strides tuple the launcher can use to build the
 permuted-B torch tensor).

Pairs with the FP8 / BF8 MFMA atoms in :mod:`rocke.helpers.atoms`:
the per-lane operand width (``a_per_lane`` / ``b_per_lane`` = 8 fp8 /
bf8 bytes = ``<2 x i32>``) determines the contiguous-load size the
preshuffle layout has to package.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Tuple

from ..core.ir import IRBuilder, Value


__all__ = [
    "PreshuffleBSpec",
    "emit_preshuffleb_offset",
    "host_preshuffle_layout",
]


@dataclass(frozen=True)
class PreshuffleBSpec:
    """One concrete preshuffled-B tile shape.

    ``block_n`` / ``block_k`` are the GEMM block tile sizes. The
    preshuffled layout packs each ``(block_n, block_k)`` tile
    contiguously, so the per-block byte stride is
    ``block_n * block_k * elem_bytes``.
    """

    block_n: int
    block_k: int
    elem_bytes: int = (
        1  # 1 for fp8/bf8/i8; 2 for f16/bf16; 0.5 for i4 (use 1 with 2-per-byte packing)
    )

    @property
    def tile_bytes(self) -> int:
        """``block_n * block_k * elem_bytes`` -- bytes per preshuffled tile."""
        return self.block_n * self.block_k * self.elem_bytes


def emit_preshuffleb_offset(
    b: IRBuilder,
    spec: PreshuffleBSpec,
    *,
    n_tile: Value,
    k_tile: Value,
    n_in_tile: Value,
    k_in_tile: Value,
    n_tile_count: Value,
) -> Value:
    """Per-lane byte offset for one ``(n_tile, k_tile, n_in_tile, k_in_tile)`` quad.

    Layout (matches CK Tile's preshuffle-B writer):

    .. code-block:: text

    offset = (k_tile * n_tile_count + n_tile) * tile_bytes
    + (n_in_tile * block_k + k_in_tile) * elem_bytes

    The outer dim is K (so consecutive K-tiles in the K-loop touch
    contiguous tiles), then N, then the per-tile (N, K) layout. This
    is the canonical order CK Tile's ``permute_b`` writer produces.

    ``n_tile_count`` is the number of N-tiles per block-N stripe
    (i.e., ``ceil(N / block_n)`` in the standard non-stripe case;
    larger when the spec's stripe size differs).
    """
    c_tile_bytes = b.const_i32(spec.tile_bytes)
    c_block_k = b.const_i32(spec.block_k)
    c_elem_bytes = b.const_i32(spec.elem_bytes)

    tile_id = b.add(b.mul(k_tile, n_tile_count), n_tile)
    tile_base = b.mul(tile_id, c_tile_bytes)
    inner = b.add(b.mul(n_in_tile, c_block_k), k_in_tile)
    inner_bytes = b.mul(inner, c_elem_bytes)
    return b.add(tile_base, inner_bytes)


def host_preshuffle_layout(
    spec: PreshuffleBSpec, *, n: int, k: int
) -> Tuple[Tuple[int, ...], Tuple[int, ...]]:
    """Pure-Python layout descriptor for the host-side permutation.

    Returns ``(shape, strides)`` for a torch / numpy view of the
    preshuffled buffer::

    shape = (k_tiles, n_tiles, block_n, block_k)
    strides = (n_tiles * block_n * block_k,
    block_n * block_k,
    block_k,
    1)

    The launcher uses this to build a contiguous-after-permute torch
    tensor: ``b_pre = b_natural.unfold(...).permute(...).contiguous()``.
    """
    n_tiles = (n + spec.block_n - 1) // spec.block_n
    k_tiles = (k + spec.block_k - 1) // spec.block_k
    if n_tiles * spec.block_n != n or k_tiles * spec.block_k != k:
        raise ValueError(
            f"preshuffle requires N / K to divide block_n / block_k "
            f"(got N={n}, block_n={spec.block_n}, "
            f"K={k}, block_k={spec.block_k})"
        )
    shape = (k_tiles, n_tiles, spec.block_n, spec.block_k)
    strides = (
        n_tiles * spec.block_n * spec.block_k,
        spec.block_n * spec.block_k,
        spec.block_k,
        1,
    )
    return shape, strides

File: dnn-providers/hip-kernel-provider/rocke/platform/Cpp/helpers/preshuffle.cpp
// Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
// SPDX-License-Identifier: MIT
/*
 * C99 port of rocke/helpers/preshuffle.py: PreshuffleBSpec,
 * emit_preshuffleb_offset, host_preshuffle_layout.
 *
 * See the header for the original Python and the contract. The single
 * builder-emitting symbol (emit_preshuffleb_offset) reproduces the const_i32 /
 * mul / add call sequence byte-faithfully so the downstream IR is identical to
 * the Python.
 */
#include "rocke/helper_rocke.helpers.preshuffle.h"

#include "rocke/ir_internal.h" /* rocke_i_set_err */

/* ------------------------------------------------------------------ *
 * PreshuffleBSpec.tile_bytes
 * ------------------------------------------------------------------ */

int rocke_preshuffleb_spec_tile_bytes(const rocke_preshuffleb_spec_t* spec)
{
    /* Python: block_n * block_k * elem_bytes.
     * The fields are small Python ints whose product fits in 32 bits for every
     * real tile shape, so a plain int multiply reproduces the value exactly. */
    if(spec == NULL)
    {
        return 0;
    }
    return spec->block_n * spec->block_k * spec->elem_bytes;
}

/* ------------------------------------------------------------------ *
 * emit_preshuffleb_offset
 * ------------------------------------------------------------------ */

rocke_value_t* rocke_emit_preshuffleb_offset(rocke_ir_builder_t* b,
                                             const rocke_preshuffleb_spec_t* spec,
                                             rocke_value_t* n_tile,
                                             rocke_value_t* k_tile,
                                             rocke_value_t* n_in_tile,
                                             rocke_value_t* k_in_tile,
                                             rocke_value_t* n_tile_count)
{
    rocke_value_t* c_tile_bytes;
    rocke_value_t* c_block_k;
    rocke_value_t* c_elem_bytes;
    rocke_value_t* tile_id;
    rocke_value_t* tile_base;
    rocke_value_t* inner;
    rocke_value_t* inner_bytes;

    if(b != NULL && b->status != ROCKE_OK)
    {
        return NULL; /* already in error: no-op */
    }
    if(spec == NULL)
    {
        rocke_i_set_err(b, ROCKE_ERR_VALUE, "emit_preshuffleb_offset: NULL spec");
        return NULL;
    }

    /* Python, in this exact order:
     *   c_tile_bytes = b.const_i32(spec.tile_bytes)
     *   c_block_k    = b.const_i32(spec.block_k)
     *   c_elem_bytes = b.const_i32(spec.elem_bytes)
     *   tile_id     = b.add(b.mul(k_tile, n_tile_count), n_tile)
     *   tile_base   = b.mul(tile_id, c_tile_bytes)
     *   inner       = b.add(b.mul(n_in_tile, c_block_k), k_in_tile)
     *   inner_bytes = b.mul(inner, c_elem_bytes)
     *   return b.add(tile_base, inner_bytes)
     */
    c_tile_bytes = rocke_b_const_i32(b, rocke_preshuffleb_spec_tile_bytes(spec));
    c_block_k = rocke_b_const_i32(b, spec->block_k);
    c_elem_bytes = rocke_b_const_i32(b, spec->elem_bytes);

    tile_id = rocke_b_add(b, rocke_b_mul(b, k_tile, n_tile_count), n_tile);
    tile_base = rocke_b_mul(b, tile_id, c_tile_bytes);
    inner = rocke_b_add(b, rocke_b_mul(b, n_in_tile, c_block_k), k_in_tile);
    inner_bytes = rocke_b_mul(b, inner, c_elem_bytes);
    return rocke_b_add(b, tile_base, inner_bytes);
}

/* ------------------------------------------------------------------ *
 * host_preshuffle_layout
 * ------------------------------------------------------------------ */

rocke_status_t rocke_host_preshuffle_layout(
    const rocke_preshuffleb_spec_t* spec, int n, int k, int out_shape[4], int out_strides[4])
{
    int n_tiles;
    int k_tiles;

    if(spec == NULL)
    {
        return ROCKE_ERR_VALUE;
    }

    /* Python: floor-division on positive ints == ceil((n)/block) given the
     * +block-1 numerator. The real specs always have block_n / block_k > 0. */
    n_tiles = (n + spec->block_n - 1) / spec->block_n;
    k_tiles = (k + spec->block_k - 1) / spec->block_k;

    /* Python:
     *   if n_tiles*block_n != n or k_tiles*block_k != k: raise ValueError(...)
     */
    if(n_tiles * spec->block_n != n || k_tiles * spec->block_k != k)
    {
        return ROCKE_ERR_VALUE;
    }

    /* shape   = (k_tiles, n_tiles, block_n, block_k) */
    if(out_shape != NULL)
    {
        out_shape[0] = k_tiles;
        out_shape[1] = n_tiles;
        out_shape[2] = spec->block_n;
        out_shape[3] = spec->block_k;
    }
    /* strides = (n_tiles*block_n*block_k, block_n*block_k, block_k, 1) */
    if(out_strides != NULL)
    {
        out_strides[0] = n_tiles * spec->block_n * spec->block_k;
        out_strides[1] = spec->block_n * spec->block_k;
        out_strides[2] = spec->block_k;
        out_strides[3] = 1;
    }
    return ROCKE_OK;
}

rocke_status_t rocke_b_host_preshuffle_layout(rocke_ir_builder_t* b,
                                              const rocke_preshuffleb_spec_t* spec,
                                              int n,
                                              int k,
                                              int out_shape[4],
                                              int out_strides[4])
{
    rocke_status_t st;

    if(b != NULL && b->status != ROCKE_OK)
    {
        return b->status; /* already in error: no-op */
    }
    if(spec == NULL)
    {
        rocke_i_set_err(b, ROCKE_ERR_VALUE, "host_preshuffle_layout: NULL spec");
        return ROCKE_ERR_VALUE;
    }

    st = rocke_host_preshuffle_layout(spec, n, k, out_shape, out_strides);
    if(st != ROCKE_OK)
    {
        /* Python:
         *   raise ValueError(
         *       "preshuffle requires N / K to divide block_n / block_k "
         *       "(got N=..., block_n=..., K=..., block_k=...)") */
        rocke_i_set_err(b,
                        ROCKE_ERR_VALUE,
                        "preshuffle requires N / K to divide block_n / block_k "
                        "(got N=%d, block_n=%d, K=%d, block_k=%d)",
                        n,
                        spec->block_n,
                        k,
                        spec->block_k);
        return st;
    }
    return ROCKE_OK;
}
